File: javascript/05-best_time_to_buy_sell.js
// You are given an array prices where prices[i] is the price of a given stock on the ith day.

// You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

// Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

// Example 1:

// Input: prices = [7,1,5,3,6,4]
// Output: 5
// Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
// Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
// Example 2:

// Input: prices = [7,6,4,3,1]
// Output: 0
// Explanation: In this case, no transactions are done and the max profit = 0.

// using bruce forte approach

const maxProfit = (prices) => {
  let largestProfit = 0;

  for (let i = 0; i < prices.length; i++) {
    for (let j = 1; j < prices.length; j++) {
      let currentProfit = prices[j] - prices[i];
      if (currentProfit > largestProfit) {
        largestProfit = currentProfit;
      }
    }
  }
  return largestProfit;
};

const maxProfit2 = (nums) => {
  let left = 0;
  let right = 1;
  let max_profit = 0;
  while (right < nums.length) {
    if (nums[left] < nums[right]) {
      let currentProfit = nums[right] - nums[left];
      max_profit = Math.max(currentProfit, max_profit);
    } else {
      left = right;
    }
    right++;
  }
  return max_profit;
};

const maxProfit3 = (prices) => {
  min_price = prices[0];
  largestProfit = 0;
  for (let price of prices) {
    if (price < min_price) {
      min_price = price;
    } else if (price - min_price > largestProfit) {
      largestProfit = price - min_price;
    }
  }
  return largestProfit;
};

nums = [7, 1, 5, 3, 6, 4];
console.log(maxProfit(nums));

File: python/05-best_time_to_buy_sell.py
# You are given an array prices where prices[i] is the price of a given stock on the ith day.

# You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

# Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.


# Example 1:

# Input: prices = [7,1,5,3,6,4]
# Output: 5
# Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
# Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
# Example 2:

# Input: prices = [7,6,4,3,1]
# Output: 0
# Explanation: In this case, no transactions are done and the max profit = 0.

# Usning Bruce forte Approach
def maxProfit(nums):
    largestProfit = 0
    for i in range(0, len(nums)):
        for j in range(1, len(nums)):
            currentProfit = nums[j] - nums[i]
            if (currentProfit > largestProfit):
                largestProfit = currentProfit
    return largestProfit


# Using two  pointer approach with builtin max function
def maxProfit2(nums):
    left = 0
    right = 1
    max_profit = 0
    while right < len(nums):
        if nums[left] < nums[right]:
            currentProfit = nums[right]-nums[left]
            max_profit = max(currentProfit, max_profit)
        else:
            left = right
        right += 1
    return max_profit

# Using two pointer approach


def maxProfit3(nums):
    min_price = nums[0]
    max_profit = 0
    for i in range(len(nums)):
        if (nums[i] < min_price):
            min_price = nums[i]
        elif (nums[i]-min_price > max_profit):
            max_profit = nums[i]-min_price
    return max_profit


nums = [7, 1, 5, 3, 6, 4]
print(maxProfit3(nums))
